### PR TITLE
evoting: use skipchain service in protocol

### PR DESCRIPTION
**Pull request by gnarula**
Optimisations for Mix and Partial operations used within shuffle and decrypt protocol. They now use the skipchain service to get the blocks. The iterations are also performed from the end to search for the Mix or Partial transactions since they're always at the end of the chain in case of a finished election.

### CODE REVIEW

File: evoting/lib/election.go [hunk 1]
```diff
@@ -136,10 +136,14 @@ func (e *Election) setStage(s *skipchain.Service) error {
 }
 
 // Box accumulates all the ballots while only keeping the last ballot for each user.
-func (e *Election) Box() (*Box, error) {
-	client := skipchain.NewClient()
+func (e *Election) Box(s *skipchain.Service) (*Box, error) {
+	// client := skipchain.NewClient()
```
Comment (ineiti) [inline evoting/lib/election.go]: remove unused code

File: evoting/lib/election.go [hunk 2]
```diff
@@ -181,54 +188,77 @@ func (e *Election) Box() (*Box, error) {
 }
 
 // Mixes returns all mixes created by the roster conodes.
-func (e *Election) Mixes() ([]*Mix, error) {
-	client := skipchain.NewClient()
+func (e *Election) Mixes(s *skipchain.Service) ([]*Mix, error) {
 
-	block, err := client.GetSingleBlockByIndex(e.Roster, e.ID, 0)
+	// Traversing forward one by one might be expensive in a large
+	// election. Better would be to search for the Mix transactions
```
Comment (ineiti) [inline evoting/lib/election.go]: s/would be/is/

File: evoting/lib/election.go [hunk 3]
```diff
@@ -181,54 +188,77 @@ func (e *Election) Box() (*Box, error) {
...
 }
 
 // Partials returns the partial decryption for each roster conode.
-func (e *Election) Partials() ([]*Partial, error) {
-	client := skipchain.NewClient()
+func (e *Election) Partials(s *skipchain.Service) ([]*Partial, error) {
 
-	block, err := client.GetSingleBlockByIndex(e.Roster, e.ID, 0)
+	// Traversing forward one by one might be expensive in a large
+	// election. Better would be to search for the Partial transactions
```
Comment (ineiti) [inline evoting/lib/election.go]: s/Better would be/It is better/

File: evoting/lib/chains.go [hunk 4]
```diff
@@ -32,17 +32,15 @@ func NewSkipchain(s *skipchain.Service, roster *onet.Roster, verifier []skipchai
 // using websockets. Used for storing a block while executing a protocol.
 func StoreUsingWebsocket(id skipchain.SkipBlockID, roster *onet.Roster, transaction *Transaction) error {
 	client := skipchain.NewClient()
-	reply, err := client.GetUpdateChain(roster, id)
+	latest, err := client.GetSingleBlockByIndex(roster, id, -1)
```
Comment (ineiti) [inline evoting/lib/chains.go]: I think we should not use the `-1` and remove it. It opens up too big a hole for the server to give wrong blocks. `GetUpdateChain` is much more secure, as it checks the proof of the blocks.

**Comment by gnarula** [comment]
Addressed the comments in the new commit

File: evoting/lib/election.go [hunk 5]
```diff
@@ -181,54 +186,77 @@ func (e *Election) Box() (*Box, error) {
...
+
+		if transaction.Mix != nil {
+			// append to the mixes array
 			mixes = append(mixes, transaction.Mix)
 		}
 
-		if len(block.ForwardLink) <= 0 {
-			break
+		// keep iterating back
+		block, err = s.GetSingleBlock(&skipchain.GetSingleBlock{ID: block.BackLinkIDs[0]})
```
Comment (ineiti) [inline evoting/lib/election.go]: Why don't you use `s.GetDB().GetByID` here?

File: evoting/lib/election.go [hunk 6]
```diff
@@ -181,54 +186,77 @@ func (e *Election) Box() (*Box, error) {
...
-		if len(block.ForwardLink) <= 0 {
-			break
+		if transaction.Partial != nil {
+			partials = append(partials, transaction.Partial)
 		}
-		var err error
-		block, err = client.GetSingleBlock(e.Roster, block.ForwardLink[0].To)
+
+		// keep iterating back
+		block, err = s.GetSingleBlock(&skipchain.GetSingleBlock{ID: block.BackLinkIDs[0]})
```
Comment (ineiti) [inline evoting/lib/election.go]: Why not `s.GetDB().GetByID`?

**Comment by ineiti** [comment]
Two more comments, but then it's OK.

File: evoting/lib/election.go (hunk 5 above)
Comment (gnarula) [reply evoting/lib/election.go]: Fixed

File: evoting/lib/election.go (hunk 6 above)
Comment (gnarula) [reply evoting/lib/election.go]: Fixed

**Pull request merged**